### ISSUE TITLE
MyArrayListMoreOps.doMultiple() is prone to dead-code elimination

### ISSUE THREAD

**Issue by vhotspur** [arahusky/performance_javadoc#23]
The result of the `contains` operation is [nowhere stored](https://github.com/arahusky/performance_javadoc/blob/2244139d5941553e276fade90a20dd84992e1864/src/java-demo/example001/MyArrayListMoreOps.java#L53) and thus the JIT compiler will eliminate the whole loop.

In the iteration loop, the compiler may detect that the result (`o` variable) is nowhere used and thus can be discarded. This may actually lead to removal of the whole loop (both inner and outer one).

The remedy is not easy. Returning `int` that would combine hash codes (probably `System.identityHashCode` rather than possibly custom hash code of the element) might help. See the `BlackHole` class in [jmh](http://openjdk.java.net/projects/code-tools/jmh/). The trick with returning of the value would need some support in the harness as well.

**Comment by vhotspur** [comment]
The compiler may still choose to optimize-out this loop:

```java
for (int i = 0; i < searches; i++) {
    this.contains(i);
}
```

`contains()` probably does not have any side-effect and thus can be safely removed. The solution is to use blackhole here or xor with something that is being returned.

The framework shall offer some decent out-of-the-box support for this anyway.

**Comment by vhotspur** [comment]
Could you, please, re-open this? Thanks :-)

**Comment by arahusky** [comment]
This should be fixed by #ff4181b7786f3321fb49e791104d3409fbba4728.

**Comment by arahusky** [comment]
This should be finally fixed by ff4181b7786f3321fb49e791104d3409fbba4728

**Comment by arahusky** [comment]
This should be finally fixed by ff4181b7786f3321fb49e791104d3409fbba4728
